File: sdk/elixir/lib/fountain/turn_follower.ex
defmodule Fountain.TurnFollower do
  @moduledoc "Folds a raw conversation event feed into one turn's answer."
  defstruct [
    :turn_number,
    :turn_id,
    :state,
    :exit_code,
    :reason,
    started: false,
    chunks: [],
    tools: [],
    break_before_text: false
  ]

  def new(turn_number, turn_id \\ nil),
    do: %__MODULE__{turn_number: turn_number, turn_id: turn_id}

  def text(value), do: value.chunks |> Enum.reverse() |> IO.iodata_to_binary() |> String.trim()
  def tools_used(value), do: Enum.reverse(value.tools)
  def finished?(value), do: not is_nil(value.state)

  @doc "Applies one raw, string-keyed API event, returning `{follower, run_events}`."
  def apply(follower, %{"kind" => "stage"} = event), do: apply_stage(follower, event)
  def apply(follower, %{"kind" => "output"} = event), do: apply_output(follower, event)
  def apply(follower, _event), do: {follower, []}

  defp apply_stage(follower, %{"stage" => "turn"} = event) do
    meta = object(event["data"])

    if matches?(follower, meta) do
      case event["state"] do
        "started" ->
          follower = %{
            follower
            | started: true,
              turn_id: string(meta["turn_id"]) || follower.turn_id
          }

          {follower,
           [%{type: :turn_start, turn_number: follower.turn_number, turn_id: follower.turn_id}]}

        state when state in ~w(done failed interrupted) ->
          follower = %{
            follower
            | state: String.to_atom(state),
              turn_id: follower.turn_id || string(meta["turn_id"]),
              exit_code:
                if(is_integer(meta["exit_code"]), do: meta["exit_code"], else: follower.exit_code),
              reason: string(meta["reason"]) || string(meta["stop_reason"]) || follower.reason
          }

          {follower,
           [
             %{
               type: :turn_end,
               state: follower.state,
               exit_code: follower.exit_code,
               reason: follower.reason
             }
           ]}

        _ ->
          {follower, []}
      end
    else
      {follower, []}
    end
  end

  defp apply_stage(follower, _event), do: {follower, []}

  defp apply_output(follower, event) do
    event_turn_id = string(event["turn_id"])

    if (not is_nil(follower.turn_id) and not is_nil(event_turn_id) and
          event_turn_id != follower.turn_id) or
         (not follower.started and is_nil(follower.turn_id)) do
      {follower, []}
    else
      Enum.reduce(event["blocks"] || [], {follower, []}, fn block, {state, output} ->
        if is_map(block) do
          {state, derived} = apply_block(state, block, event["stream"] == "acp")
          {state, output ++ [%{type: :block, block: block, event: event}] ++ derived}
        else
          {state, output}
        end
      end)
    end
  end

  defp apply_block(follower, %{"kind" => "text"} = block, acp) do
    body = if is_binary(block["body"]), do: block["body"], else: ""
    if body == "", do: {follower, []}, else: append_text(follower, body, acp)
  end

  defp apply_block(follower, %{"kind" => "thinking"} = block, _),
    do:
      {follower,
       if(is_binary(block["body"]) and block["body"] != "",
         do: [%{type: :thinking, text: block["body"]}],
         else: []
       )}

  defp apply_block(follower, %{"kind" => kind}, _) when kind in ["raw", "init"],
    do: {follower, []}

  defp apply_block(follower, %{"kind" => "permission_request"} = block, _) do
    follower = %{follower | break_before_text: true}

    case permission_request(block) do
      nil -> {follower, []}
      request -> {follower, [%{type: :permission, request: request, block: block}]}
    end
  end

  defp apply_block(follower, %{"kind" => "tool_use"} = block, _) do
    follower = %{follower | break_before_text: true}
    name = string(block["name"])

    follower =
      if name && name not in follower.tools,
        do: %{follower | tools: [name | follower.tools]},
        else: follower

    {follower, if(name, do: [%{type: :tool, name: name, block: block}], else: [])}
  end

  defp apply_block(%{chunks: []} = follower, %{"kind" => "result", "body" => body}, _)
       when is_binary(body) and body != "",
       do: {%{follower | chunks: [body], break_before_text: true}, [%{type: :text, text: body}]}

  defp apply_block(follower, %{"kind" => "error", "body" => body}, _)
       when is_binary(body) and body != "" do
    text = "\n[error] #{body}\n"

    {%{follower | chunks: [text | follower.chunks], break_before_text: true},
     [%{type: :text, text: text}]}
  end

  defp apply_block(follower, _block, _), do: {%{follower | break_before_text: true}, []}

  defp append_text(follower, body, acp) do
    prefix =
      cond do
        follower.chunks == [] -> ""
        acp and not follower.break_before_text -> ""
        String.ends_with?(hd(follower.chunks), "\n") -> ""
        true -> "\n\n"
      end

    chunks = if prefix == "", do: [body | follower.chunks], else: [body, prefix | follower.chunks]

    {%{follower | chunks: chunks, break_before_text: false},
     [%{type: :text, text: prefix <> body}]}
  end

  defp permission_request(block) do
    request_id = string(block["request_id"])

    options =
      (block["options"] || [])
      |> Enum.flat_map(fn option ->
        if is_map(option) do
          id = string(option["optionId"]) || string(option["option_id"])
          if id, do: [Map.put(option, :option_id, id)], else: []
        else
          []
        end
      end)

    if request_id && options != [],
      do: %{
        request_id: request_id,
        summary: string(block["summary"]) || string(block["body"]),
        tool_name: string(block["name"]),
        tool_id: string(block["tool_id"]),
        options: options
      }
  end

  defp matches?(%{turn_id: id} = follower, meta) when is_binary(id),
    do: string(meta["turn_id"]) == id or meta["turn_number"] == follower.turn_number

  defp matches?(follower, meta), do: meta["turn_number"] == follower.turn_number
  defp object(value) when is_map(value), do: value

  defp object(value) when is_binary(value) do
    case Jason.decode(value) do
      {:ok, map} when is_map(map) -> map
      _ -> %{}
    end
  end

  defp object(_), do: %{}
  defp string(value) when is_binary(value) and value != "", do: value
  defp string(_), do: nil
end

File: sdk/swift/Sources/FountainKit/SSE/TurnFollower.swift
import Foundation

/// What a turn produces, distilled: streamable pieces plus the final state.
public enum TurnEvent: Sendable {
  /// The conversation the turn belongs to, before it starts — `Run` emits
  /// this first so a caller can link or title the transcript immediately.
  /// `TurnFollower` never emits it; it follows a turn, not a conversation.
  case conversation(Conversation, url: URL)
  case turnStart(turnNumber: Int, turnID: String?)
  /// A chunk of the answer, with paragraph breaks already applied.
  case text(String)
  case thinking(String)
  case tool(name: String, block: Block)
  case permission(PermissionRequest, block: Block)
  /// Every block, before interpretation — for transcript views.
  case block(Block, event: LogEvent)
  case turnEnd(state: TurnState, exitCode: Int?, reason: String?)
}

/// How a followed turn ended. `timeout` is client-side only.
public struct TurnState: WireValue {
  public let rawValue: String
  public init(rawValue: String) { self.rawValue = rawValue }

  public static let done: Self = "done"
  public static let failed: Self = "failed"
  public static let interrupted: Self = "interrupted"
  public static let timeout: Self = "timeout"
}

/// Folds a multi-turn, multi-stream log feed into one turn's answer.
/// Port of the TypeScript SDK's `TurnFollower` — the semantics (turn
/// matching, paragraph joining, which block kinds count as the answer)
/// are deliberately identical.
public struct TurnFollower: Sendable {
  public let turnNumber: Int
  public private(set) var turnID: String?
  public private(set) var started = false
  public private(set) var state: TurnState?
  public private(set) var exitCode: Int?
  public private(set) var reason: String?
  public private(set) var tools: [String] = []

  private var chunks: [String] = []
  private var breakBeforeText = false

  public init(turnNumber: Int) {
    self.turnNumber = turnNumber
  }

  /// The accumulated answer so far.
  public var text: String {
    chunks.joined().trimmingCharacters(in: .whitespacesAndNewlines)
  }

  public var finished: Bool { state != nil }

  public mutating func apply(_ event: LogEvent) -> [TurnEvent] {
    switch event.kind {
    case .stage: applyStage(event)
    case .output: applyOutput(event)
    default: []
    }
  }

  private mutating func applyStage(_ event: LogEvent) -> [TurnEvent] {
    guard event.stage == "turn", let meta = event.stageData else { return [] }
    let metaTurnID = meta["turn_id"]?.stringValue
    let metaTurnNumber: Int? =
      if case .number(let n) = meta["turn_number"] { Int(n) } else { nil }

    // Match on turn id when we know it, else on turn number.
    let matches =
      if let turnID, let metaTurnID { metaTurnID == turnID } else { metaTurnNumber == turnNumber }
    guard matches else { return [] }

    switch event.state {
    case .started:
      started = true
      if turnID == nil { turnID = metaTurnID }
      return [.turnStart(turnNumber: turnNumber, turnID: turnID)]
    case .done, .failed, .interrupted:
      let ended = TurnState(rawValue: event.state!.rawValue)
      state = ended
      if case .number(let code) = meta["exit_code"] { exitCode = Int(code) }
      reason = meta["reason"]?.stringValue ?? meta["stop_reason"]?.stringValue
      return [.turnEnd(state: ended, exitCode: exitCode, reason: reason)]
    default:
      return []
    }
  }

  private mutating func applyOutput(_ event: LogEvent) -> [TurnEvent] {
    // Someone else's turn (or history replay of an older one).
    if let eventTurn = event.turnID, let turnID, eventTurn != turnID { return [] }
    // Tail of an older turn arriving before ours starts.
    if !started && turnID == nil { return [] }

    let acp = event.stream == .acp
    var events: [TurnEvent] = []
    for block in event.blocks ?? [] {
      events.append(.block(block, event: event))
      events.append(contentsOf: apply(block: block, acp: acp))
    }
    return events
  }

  private mutating func apply(block: Block, acp: Bool) -> [TurnEvent] {
    switch block.kind {
    case .text:
      guard let body = block.body, !body.isEmpty else { return [] }
      let prefix = paragraphBreak(acp: acp)
      chunks.append(prefix + body)
      breakBeforeText = false
      return [.text(prefix + body)]
    case .thinking:
      guard let body = block.body else { return [] }
      return [.thinking(body)]
    case .raw, .initialize:
      // Transport bookkeeping; not even a paragraph break.
      return []
    case .permissionRequest:
      breakBeforeText = true
      guard let request = PermissionRequest(block: block) else { return [] }
      return [.permission(request, block: block)]
    case .toolUse:
      breakBeforeText = true
      guard let name = block.name else { return [] }
      if !tools.contains(name) { tools.append(name) }
      return [.tool(name: name, block: block)]
    case .result:
      breakBeforeText = true
      // Counts as the answer only when nothing else did.
      guard chunks.isEmpty, let body = block.body, !body.isEmpty else { return [] }
      chunks.append(body)
      return [.text(body)]
    case .error:
      breakBeforeText = true
      guard let body = block.body else { return [] }
      let line = "\n[error] \(body)\n"
      chunks.append(line)
      return [.text(line)]
    default:
      breakBeforeText = true
      return []
    }
  }

  /// ACP chunks are pieces of one message and join with nothing; legacy
  /// stdout rows are whole messages and join as paragraphs; text after a
  /// tool call always starts a new paragraph.
  private func paragraphBreak(acp: Bool) -> String {
    guard let last = chunks.last else { return "" }
    if acp && !breakBeforeText { return "" }
    return last.hasSuffix("\n") ? "" : "\n\n"
  }
}
